https://adventofcode.com/2021/day/4


<h2>--- Day 4: Giant Squid ---</h2><p>You're already almost 1.5km (almost a mile) below the surface of the ocean, already so deep that you can't see any sunlight. What you <em>can</em> see, however, is a giant squid that has attached itself to the outside of your submarine.</p>
<p>Maybe it wants to play <a href="https://en.wikipedia.org/wiki/Bingo_(American_version)" target="_blank">bingo</a>?</p>
<p>Bingo is played on a set of boards each consisting of a 5x5 grid of numbers. Numbers are chosen at random, and the chosen number is <em>marked</em> on all boards on which it appears. (Numbers may not appear on all boards.) If all numbers in any row or any column of a board are marked, that board <em>wins</em>. (Diagonals don't count.)</p>
<p>The submarine has a <em>bingo subsystem</em> to help passengers (currently, you and the giant squid) pass the time. It automatically generates a random order in which to draw numbers and a random set of boards (your puzzle input). For example:</p>
<pre><code>7,4,9,5,11,17,23,2,0,14,21,24,10,16,13,6,15,25,12,22,18,20,8,19,3,26,1

22 13 17 11  0
 8  2 23  4 24
21  9 14 16  7
 6 10  3 18  5
 1 12 20 15 19

 3 15  0  2 22
 9 18 13 17  5
19  8  7 25 23
20 11 10 24  4
14 21 16 12  6

14 21 17 24  4
10 16 15  9 19
18  8 23 26 20
22 11 13  6  5
 2  0 12  3  7
</code></pre>
<p>After the first five numbers are drawn (<code>7</code>, <code>4</code>, <code>9</code>, <code>5</code>, and <code>11</code>), there are no winners, but the boards are marked as follows (shown here adjacent to each other to save space):</p>
<pre><code>22 13 17 <em>11</em>  0         3 15  0  2 22        14 21 17 24  <em>4</em>
 8  2 23  <em>4</em> 24         <em>9</em> 18 13 17  <em>5</em>        10 16 15  <em>9</em> 19
21  <em>9</em> 14 16  <em>7</em>        19  8  <em>7</em> 25 23        18  8 23 26 20
 6 10  3 18  <em>5</em>        20 <em>11</em> 10 24  <em>4</em>        22 <em>11</em> 13  6  <em>5</em>
 1 12 20 15 19        14 21 16 12  6         2  0 12  3  <em>7</em>
</code></pre>
<p>After the next six numbers are drawn (<code>17</code>, <code>23</code>, <code>2</code>, <code>0</code>, <code>14</code>, and <code>21</code>), there are still no winners:</p>
<pre><code>22 13 <em>17</em> <em>11</em>  <em>0</em>         3 15  <em>0</em>  <em>2</em> 22        <em>14</em> <em>21</em> <em>17</em> 24  <em>4</em>
 8  <em>2</em> <em>23</em>  <em>4</em> 24         <em>9</em> 18 13 <em>17</em>  <em>5</em>        10 16 15  <em>9</em> 19
<em>21</em>  <em>9</em> <em>14</em> 16  <em>7</em>        19  8  <em>7</em> 25 <em>23</em>        18  8 <em>23</em> 26 20
 6 10  3 18  <em>5</em>        20 <em>11</em> 10 24  <em>4</em>        22 <em>11</em> 13  6  <em>5</em>
 1 12 20 15 19        <em>14</em> <em>21</em> 16 12  6         <em>2</em>  <em>0</em> 12  3  <em>7</em>
</code></pre>
<p>Finally, <code>24</code> is drawn:</p>
<pre><code>22 13 <em>17</em> <em>11</em>  <em>0</em>         3 15  <em>0</em>  <em>2</em> 22        <em>14</em> <em>21</em> <em>17</em> <em>24</em>  <em>4</em>
 8  <em>2</em> <em>23</em>  <em>4</em> <em>24</em>         <em>9</em> 18 13 <em>17</em>  <em>5</em>        10 16 15  <em>9</em> 19
<em>21</em>  <em>9</em> <em>14</em> 16  <em>7</em>        19  8  <em>7</em> 25 <em>23</em>        18  8 <em>23</em> 26 20
 6 10  3 18  <em>5</em>        20 <em>11</em> 10 <em>24</em>  <em>4</em>        22 <em>11</em> 13  6  <em>5</em>
 1 12 20 15 19        <em>14</em> <em>21</em> 16 12  6         <em>2</em>  <em>0</em> 12  3  <em>7</em>
</code></pre>
<p>At this point, the third board <em>wins</em> because it has at least one complete row or column of marked numbers (in this case, the entire top row is marked: <code><em>14 21 17 24  4</em></code>).</p>
<p>The <em>score</em> of the winning board can now be calculated. Start by finding the <em>sum of all unmarked numbers</em> on that board; in this case, the sum is <code>188</code>. Then, multiply that sum by <em>the number that was just called</em> when the board won, <code>24</code>, to get the final score, <code>188 * 24 = <em>4512</em></code>.</p>
<p>To guarantee victory against the giant squid, figure out which board will win first. <em>What will your final score be if you choose that board?</em></p>


In [4]:
import pandas as pd
import numpy as np

In [5]:
f = open("bingo.txt", "r")
raw = f.read()

In [6]:
l = raw.split("\n") 
l[:4]

['31,88,35,24,46,48,95,42,18,43,71,32,92,62,97,63,50,2,60,58,74,66,15,87,57,34,14,3,54,93,75,22,45,10,56,12,83,30,8,76,1,78,82,39,98,37,19,26,81,64,55,41,16,4,72,5,52,80,84,67,21,86,23,91,0,68,36,13,44,20,69,40,90,96,27,77,38,49,94,47,9,65,28,59,79,6,29,61,53,11,17,73,99,25,89,51,7,33,85,70',
 '',
 '50 83  3 31 16',
 '47  9 94 10 86']

In [7]:
lista_sorteo = np.fromstring(l[0], sep = ',', dtype= int)
lista_sorteo[:10]

array([31, 88, 35, 24, 46, 48, 95, 42, 18, 43])

In [10]:
tableros_raw = l[1:]
tableros_raw[:15]

['',
 '50 83  3 31 16',
 '47  9 94 10 86',
 '61 22 53 46 74',
 '77 41 79 55 62',
 '97 78 43 73 40',
 '',
 '99 96 20 35 21',
 '38 17 48 69 68',
 ' 9 51 32 52 11',
 '67  8 42 89 27',
 '39 62 66 72 43',
 '',
 '33 16  4 78 31',
 '96 66 13 55 18']

In [11]:
while '' in tableros_raw:
    tableros_raw.remove('')
tableros_raw[:5]

['50 83  3 31 16',
 '47  9 94 10 86',
 '61 22 53 46 74',
 '77 41 79 55 62',
 '97 78 43 73 40']

In [14]:
def creandoListaTableros(tableros_raw):
    lista_tableros = []
    numero_t = int(len(tableros_raw)/5)
    numero_t

    seccion = 0
    for i in range(numero_t):
        t1 = np.empty((0, 5), int)
        for fila_raw in tableros_raw[seccion:seccion+5]:
            t1=np.append(t1, [np.fromstring(fila_raw, sep = ' ', dtype= int)], axis=0)
        lista_tableros.append((t1, np.full((5, 5), False, dtype=bool)))
        seccion+=5
    return lista_tableros

In [17]:
lista_tableros=creandoListaTableros(tableros_raw)
lista_tableros[:2]

[(array([[50, 83,  3, 31, 16],
         [47,  9, 94, 10, 86],
         [61, 22, 53, 46, 74],
         [77, 41, 79, 55, 62],
         [97, 78, 43, 73, 40]]),
  array([[False, False, False, False, False],
         [False, False, False, False, False],
         [False, False, False, False, False],
         [False, False, False, False, False],
         [False, False, False, False, False]])),
 (array([[99, 96, 20, 35, 21],
         [38, 17, 48, 69, 68],
         [ 9, 51, 32, 52, 11],
         [67,  8, 42, 89, 27],
         [39, 62, 66, 72, 43]]),
  array([[False, False, False, False, False],
         [False, False, False, False, False],
         [False, False, False, False, False],
         [False, False, False, False, False],
         [False, False, False, False, False]]))]

In [18]:
def tableroGanadorColumnaOFila(tablero):
    for i in range(tablero.shape[0]):
        fila = tablero[i]
        columna = tablero[:,i]
        if not False in fila:
            return True
        if not False in columna:
            return True
    return False

In [19]:
def marcarTablero(tablero, tablero_m, valor):
    for i in range(5):
        for j in range(5):
            if tablero[i][j] == valor:
                tablero_m[i][j] = True
                return True
    return False

In [20]:
def calcularPuntosTableros(tablero, tablero_m):
    suma = 0
    for i in range(5):
        for j in range(5):
            if not tablero_m[i][j]:
                suma+=tablero[i][j]
    return suma

In [21]:
t_ganador = None
for numero in lista_sorteo:
    for i_table in range(len(lista_tableros)):
        marcarTablero(lista_tableros[i_table][0], 
                     lista_tableros[i_table][1], numero)
        if tableroGanadorColumnaOFila(lista_tableros[i_table][1]):
            t_ganador = i_table
            break
    if t_ganador != None:
        break
t_ganador

66

In [22]:
p_tablero= calcularPuntosTableros(lista_tableros[t_ganador][0], lista_tableros[t_ganador][1])
p_tablero

1026

In [23]:
numero

66

In [24]:
p_tablero*numero

67716

### resultado 67716

--- Part Two ---

On the other hand, it might be wise to try a different strategy: let the giant squid win.

You aren't sure how many bingo boards a giant squid could play at once, so rather than waste time counting its arms, the safe thing to do is to figure out which board will win last and choose that one. That way, no matter which boards it picks, it will win for sure.

In the above example, the second board is the last to win, which happens after 13 is eventually called and its middle column is completely marked. If you were to keep playing until this point, the second board would have a sum of unmarked numbers equal to 148 for a final score of 148 * 13 = 1924.

Figure out which board will win last. Once it wins, what would its final score be?


In [27]:
lista_tableros=creandoListaTableros(tableros_raw)

In [28]:
t_ganador = None
lista_ganadores = []
for numero in lista_sorteo:
    for i_table in range(len(lista_tableros)):
        marcarTablero(lista_tableros[i_table][0], 
                     lista_tableros[i_table][1], numero)
        if tableroGanadorColumnaOFila(lista_tableros[i_table][1]):
            t_ganador = i_table
            if not t_ganador in lista_ganadores:
                lista_ganadores.append(t_ganador)
            #break
    if len(lista_tableros) == len(lista_ganadores):
        break
lista_ganadores[-1]

31

In [29]:
numero

6

In [30]:
t_ganador = lista_ganadores[-1]
t_ganador

31

In [31]:
p_tablero= calcularPuntosTableros(lista_tableros[t_ganador][0], lista_tableros[t_ganador][1])
p_tablero

305

In [32]:
numero*p_tablero

1830

### resultado 1830